Rest Client Reactive - multipart support

michalszynkiewicz opened the pull request.

michalszynkiewicz committed:
Rest Client Reactive multipart form support

michalszynkiewicz commented:
The failure is as follows and is not related to the change
```
Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M5:test (default-test) on project quarkus-integration-test-mongodb-client: There was a timeout in the fork -> [Help 1]
```

michalszynkiewicz commented:
I'm aware that this should get some documentation. But if this PR is otherwise okay, I'd rather add the documentation in a separate PR as I need to do some other stuff in the following days...

geoand commented:
I'll have a look on Monday.

We will absolutely need documentation for this before the final release

geoand commented on extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java:
```diff
@@ -871,6 +910,212 @@ A more full example of generated client (with sub-resource) can is at the bottom
 
     }
 
+    private ResultHandle createMultipartForm(MethodCreator methodCreator, ResultHandle methodParam, Type formClassType,
```
Can we get some comments on this? It's pretty large and I have no idea what it's supposed to do :)

michalszynkiewicz replied:
added `some` comments :)

geoand commented on independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/api/QuarkusRestClientProperties.java:
```diff
@@ -7,4 +7,9 @@
+    /**
+     * See {@link io.netty.handler.codec.http.multipart.HttpPostRequestEncoder.EncoderMode}, RFC1738 by default
+     */
+    public static final String MULTIPART_ENCODER_MODE = "io.quarkus.rest.client.multipart-post-encoder-mode";
```
We will definitely need to document this

michalszynkiewicz replied:
done

geoand commented on independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:
```diff
@@ -52,54 +59,87 @@ public void handle(Throwable event) {
+                        actualEntity = ClientSendRequestHandler.this.setMultipartHeadersAndPrepareBody(httpClientRequest,
+                                requestContext);
+
+                        Pipe<Buffer> pipe = actualEntity.pipe(); // Shouldn't this be called in an earlier phase ?
```
Should it? Did you copy this code from somewhere else?

michalszynkiewicz replied:
I don't know, this code is copied from vertx-web, with the comment.

geoand replied:
Okay, that's all I needed 😁

geoand commented on integration-tests/resteasy-reactive-rest-client-multipart/pom.xml:
```diff
@@ -116,6 +117,14 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>10</source>
+                    <target>10</target>
+                </configuration>
+            </plugin>
```
What's this about? Was added perhaps by IntelliJ?

michalszynkiewicz replied:
yes :(

michalszynkiewicz replied:
fixed

geoand commented on integration-tests/resteasy-reactive-rest-client-multipart/src/main/java/io/quarkus/it/rest/client/multipart/MultipartClient.java:
```diff
@@ -0,0 +1,119 @@
+
+        @FormParam("file")
+        @PartType(MediaType.APPLICATION_OCTET_STREAM)
+        public byte[] file;
```
We'll have to be very careful on how we document this and other blocking variants

michalszynkiewicz replied:
I updated the PR and, among others, added some basic documentation. Could you elaborate on ^? What do you think is worth mentioning?

geoand commented:
With the current implementation, is there any way a user could send a multipart stream without reading all bytes in memory?

michalszynkiewicz commented:
> With the current implementation, is there any way a user could send a multipart stream without reading all bytes in memory?

One could use files, see https://github.com/quarkusio/quarkus/pull/17173/files#diff-00c5b96175fbd95dec22f03f0df7e05d4369109d19a6547cf60bffffdc7625bbR82

michalszynkiewicz commented on integration-tests/resteasy-reactive-rest-client-multipart/src/main/java/io/quarkus/it/rest/client/multipart/MultipartClient.java:
```diff
@@ -0,0 +1,119 @@
+
+    class WithFileAsTextFile {
+        @FormParam("file")
+        @PartType(MediaType.APPLICATION_OCTET_STREAM)
```
fix this, this this should use TEXT_PLAIN

geoand commented:
> > With the current implementation, is there any way a user could send a multipart stream without reading all bytes in memory?
> 
> One could use files, see https://github.com/quarkusio/quarkus/pull/17173/files#diff-00c5b96175fbd95dec22f03f0df7e05d4369109d19a6547cf60bffffdc7625bbR82

Nice! 
Is `Path` also supported? Furthermore, where exactly in the code is that implemented? Just want to make sure it's actually non-blocking :)

michalszynkiewicz commented:
> > > With the current implementation, is there any way a user could send a multipart stream without reading all bytes in memory?
> > 
> > 
> > One could use files, see https://github.com/quarkusio/quarkus/pull/17173/files#diff-00c5b96175fbd95dec22f03f0df7e05d4369109d19a6547cf60bffffdc7625bbR82
> 
> Nice!
> Is `Path` also supported? Furthermore, where exactly in the code is that implemented? Just want to make sure it's actually non-blocking :)

at the moment only File is supported.

This creates a `io.vertx.ext.web.multipart.MultipartForm` with the fields from the multipart "body object": https://github.com/michalszynkiewicz/quarkus/blob/fd1aa99842331b2413ffcd6fe171d17bf863d1e6/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java#L913

Then the code [here](https://github.com/quarkusio/quarkus/pull/17173/files#diff-03d3c6adcdcc01b3a1f8daccc773adc6ee1c598afa245a445a24383a77f3b3a4R63) does the actual sending

geoand commented:
> at the moment only File is supported.

Let's support `Path` as well since we do on the server side as well.
 
> This creates a `io.vertx.ext.web.multipart.MultipartForm` with the fields from the multipart "body object": https://github.com/michalszynkiewicz/quarkus/blob/fd1aa99842331b2413ffcd6fe171d17bf863d1e6/extensions/resteasy-reactive/jaxrs-client-reactive/deployment/src/main/java/io/quarkus/jaxrs/client/reactive/deployment/JaxrsClientReactiveProcessor.java#L913
> 
> Then the code [here](https://github.com/quarkusio/quarkus/pull/17173/files#diff-03d3c6adcdcc01b3a1f8daccc773adc6ee1c598afa245a445a24383a77f3b3a4R63) does the actual sending

So IIUC, this code is lifted from the vertx-web client? If so, great!

michalszynkiewicz commented:
Yes, the code is heavily based on (and in some places simply copied from) vertx-web, as such, should be non-blocking.

I added support for `Path`, added documentation.
